src/s2geography/geoarrow.cc

```diff
@@ -672,6 +672,364 @@ void Reader::ReadGeography(const ArrowArray* array, int64_t offset,
+                              GEOARROW_DIMENSIONS_XY));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else if (geog.Polylines().size() == 1) {
+      // LineString
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_LINESTRING,
+                              GEOARROW_DIMENSIONS_XY));
+
+      const auto& poly = geog.Polylines()[0];
+
+      for (int i = 0; i < poly->num_vertices(); i++) {
+        S2LatLng ll(poly->vertex(i));
+        coords_[0] = ll.lng().degrees();
+        coords_[1] = ll.lat().degrees();
+        GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+      }
+
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else {
+      // MultiLineString
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_MULTILINESTRING,
+                              GEOARROW_DIMENSIONS_XY));
+
+      for (const auto& poly : geog.Polylines()) {
+        GEOARROW_RETURN_NOT_OK(
+            visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_LINESTRING,
+                                GEOARROW_DIMENSIONS_XY));
+
+        for (int i = 0; i < poly->num_vertices(); i++) {
+          S2LatLng ll(poly->vertex(i));
+          coords_[0] = ll.lng().degrees();
+          coords_[1] = ll.lat().degrees();
+          GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+        }
+
+        GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+      }
+
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+    }
+
+    return GEOARROW_OK;
+  }
+
+  int VisitLoopShell(const S2Loop* loop) {
+
+    if (loop->num_vertices() == 0) {
+      throw Exception("Unexpected S2Loop with 0 verties");
+    }
+
+    GEOARROW_RETURN_NOT_OK(visitor_.ring_start(&visitor_));
+
+    for (int i = 0; i <= loop->num_vertices(); i++) {
+      S2LatLng ll(loop->vertex(i));
+      coords_[0] = ll.lng().degrees();
+      coords_[1] = ll.lat().degrees();
+      GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+    }
+
+    GEOARROW_RETURN_NOT_OK(visitor_.ring_end(&visitor_));
+
+    return GEOARROW_OK;
+  }
+
+  int VisitLoopHole(const S2Loop* loop) {
+
+    if (loop->num_vertices() == 0) {
+      throw Exception("Unexpected S2Loop with 0 verties");
+    }
+
+    GEOARROW_RETURN_NOT_OK(visitor_.ring_start(&visitor_));
+
+    // For the hole, we use the vertices in reverse order to ensure the holes
+    // have the opposite orientation of the shell
+    for (int i = loop->num_vertices() - 1; i >= 0; i--) {
+      S2LatLng ll(loop->vertex(i));
+      coords_[0] = ll.lng().degrees();
+      coords_[1] = ll.lat().degrees();
+      GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+    }
+
+    S2LatLng ll(loop->vertex(loop->num_vertices() - 1));
+    coords_[0] = ll.lng().degrees();
+    coords_[1] = ll.lat().degrees();
+    GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
```

Comment:
@paleolimbot In the case of a hole, the code here (which I essentially took from the wkt writer) explicitly repeats the first coordinate pair (i.e. in the last vertex because we are iterating reversed, so we started with the last). 

But for `VisitLoopShell` above, the code does not do that (it just iterates over the vertices, and s2 doesn't duplicate the first/last).

In practice, though (at least in case of WKT), it seems both work fine, and so I assume that the visitor implemented by GeoArrowArrayWriter (or for WKT the GeoArrowWKTWriter used under the hood) checks if the first coordinate was repeated, and if not repeats it themselves? But is that meant to be a guarantee? Or should ideally the one calling the visitor ensure they are providing closed ring coordinates?

Reply:
OK, nevermind, I just noticed that in the case for the shell, we are looping until `i <= loop->num_vertices()` and not until `num_vertices() - 1`. So the `loop.vertex(i)` call wraps around, and so we are actually explicitly adding the first coordinate as well 
(that trick just not works when iterating in reverse)

Reply:
Indeed, that's the explanation I also had found